# Evolución de la natalidad y su relación con el desequilibrio en los nacimientos por sexo

Este análisis examina cómo ha evolucionado la relación entre nacimientos de niños y niñas en una muestra de 50 países, con especial atención al caso de China. Además, se analiza la evolución de la natalidad a lo largo del tiempo para identificar posibles cambios demográficos en las últimas décadas.

El indicador utilizado para analizar el desequilibrio en los nacimientos representa el número de niños nacidos por cada 100 niñas. En condiciones naturales, este valor suele situarse alrededor de 105, por lo que desviaciones respecto a este nivel pueden indicar la existencia de desequilibrios en los nacimientos.

El objetivo del análisis es identificar patrones demográficos, comparar diferentes países y analizar estas desviaciones en el contexto de la reducción de la natalidad observada en muchos países durante las últimas décadas.

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

In [2]:
# Dataset relacion entre nacimientos de niños y niñas
ruta_sexratio = r"C:\Users\GCCM_\OneDrive\Desktop\Bootcamp\proyecto\API_SP.POP.BRTH.MF_DS2_en_csv_v2_8393\API_SP.POP.BRTH.MF_DS2_en_csv_v2_8393.csv"
df_sex = pd.read_csv(ruta_sexratio, skiprows=4)
df_sex.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,Sex ratio at birth (male births per female bir...,SP.POP.BRTH.MF,1.043000,1.043000,1.043000,1.043000,1.043000,1.043000,...,1.044000,1.043000,1.044000,1.043000,1.043000,1.043000,1.043000,NaN,NaN,NaN
1,Africa Eastern and Southern,AFE,Sex ratio at birth (male births per female bir...,SP.POP.BRTH.MF,1.033654,1.033751,1.033727,1.033777,1.033825,1.033851,...,1.031747,1.032194,1.032312,1.032414,1.032396,1.032542,1.032581,NaN,NaN,NaN
2,Afghanistan,AFG,Sex ratio at birth (male births per female bir...,SP.POP.BRTH.MF,1.049000,1.049000,1.049000,1.049000,1.049000,1.049000,...,1.054000,1.053000,1.053000,1.052000,1.052000,1.052000,1.051000,NaN,NaN,NaN
3,Africa Western and Central,AFW,Sex ratio at birth (male births per female bir...,SP.POP.BRTH.MF,1.037556,1.037518,1.038030,1.038074,1.038155,1.038625,...,1.037422,1.037238,1.036761,1.036789,1.036735,1.036714,1.036765,NaN,NaN,NaN
4,Angola,AGO,Sex ratio at birth (male births per female bir...,SP.POP.BRTH.MF,1.028000,1.029000,1.029000,1.029000,1.029000,1.029000,...,1.025000,1.026000,1.026000,1.027000,1.027000,1.027000,1.027000,NaN,NaN,NaN


In [3]:
# Cargar dataset de fertilidad
ruta_fertility = r"C:\Users\GCCM_\OneDrive\Desktop\Bootcamp\proyecto\API_SP.DYN.TFRT.IN_DS2_en_csv_v2_140\API_SP.DYN.TFRT.IN_DS2_en_csv_v2_140.csv"
df_fertility = pd.read_csv(ruta_fertility, skiprows=4)
df_fertility.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2017,2018,2019,2020,2021,2022,2023,2024,2025,Unnamed: 70
0,Aruba,ABW,"Fertility rate, total (births per woman)",SP.DYN.TFRT.IN,4.567000,4.422000,4.262000,4.107000,3.940000,3.797000,...,1.785000,1.732000,1.701000,1.662000,1.631000,1.615000,1.602000,NaN,NaN,NaN
1,Africa Eastern and Southern,AFE,"Fertility rate, total (births per woman)",SP.DYN.TFRT.IN,6.650310,6.667308,6.688246,6.709226,6.724930,6.737459,...,4.569884,4.521443,4.471338,4.412973,4.350683,4.287033,4.223820,NaN,NaN,NaN
2,Afghanistan,AFG,"Fertility rate, total (births per woman)",SP.DYN.TFRT.IN,7.282000,7.284000,7.292000,7.302000,7.304000,7.305000,...,5.433000,5.327000,5.238000,5.145000,5.039000,4.932000,4.840000,NaN,NaN,NaN
3,Africa Western and Central,AFW,"Fertility rate, total (births per woman)",SP.DYN.TFRT.IN,6.468887,6.478351,6.492277,6.500229,6.516739,6.532766,...,5.098885,4.962571,4.829134,4.707399,4.637741,4.563354,4.497707,NaN,NaN,NaN
4,Angola,AGO,"Fertility rate, total (births per woman)",SP.DYN.TFRT.IN,6.708000,6.790000,6.872000,6.954000,7.036000,7.116000,...,5.600000,5.519000,5.442000,5.371000,5.304000,5.209000,5.124000,NaN,NaN,NaN


In [4]:
# Confirmar que es el indicador correcto
df_sex["Indicator Code"].unique()

array(['SP.POP.BRTH.MF'], dtype=object)

In [5]:
# Confirmar que es el indicador correcto
df_fertility["Indicator Code"].unique()

array(['SP.DYN.TFRT.IN'], dtype=object)

## Limpieza y preparación del dataset

In [6]:
# Eliminar columnas vacías tipo Unnamed
df_sex = df_sex.loc[:, ~df_sex.columns.str.contains('Unnamed')]

In [7]:
# Detectar columnas de años
years = [col for col in df_sex.columns if col.isdigit()]

In [8]:
# Seleccionar solo las columnas necesarias
df_sex = df_sex[["Country Name","Country Code"] + years]
df_sex.head()

,Country Name,Country Code,1960,1961,1962,1963,1964,1965,1966,1967,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,Aruba,ABW,1.043000,1.043000,1.043000,1.043000,1.043000,1.043000,1.042000,1.042000,...,1.044000,1.044000,1.043000,1.044000,1.043000,1.043000,1.043000,1.043000,NaN,NaN
1,Africa Eastern and Southern,AFE,1.033654,1.033751,1.033727,1.033777,1.033825,1.033851,1.033847,1.033736,...,1.031743,1.031747,1.032194,1.032312,1.032414,1.032396,1.032542,1.032581,NaN,NaN
2,Afghanistan,AFG,1.049000,1.049000,1.049000,1.049000,1.049000,1.049000,1.049000,1.049000,...,1.055000,1.054000,1.053000,1.053000,1.052000,1.052000,1.052000,1.051000,NaN,NaN
3,Africa Western and Central,AFW,1.037556,1.037518,1.038030,1.038074,1.038155,1.038625,1.039097,1.039612,...,1.037387,1.037422,1.037238,1.036761,1.036789,1.036735,1.036714,1.036765,NaN,NaN
4,Angola,AGO,1.028000,1.029000,1.029000,1.029000,1.029000,1.029000,1.029000,1.029000,...,1.025000,1.025000,1.026000,1.026000,1.027000,1.027000,1.027000,1.027000,NaN,NaN


In [9]:
df_fertility = df_fertility[["Country Name","Country Code"] + years]
df_fertility.head()

,Country Name,Country Code,1960,1961,1962,1963,1964,1965,1966,1967,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,Aruba,ABW,4.567000,4.422000,4.262000,4.107000,3.940000,3.797000,3.621000,3.452000,...,1.848000,1.785000,1.732000,1.701000,1.662000,1.631000,1.615000,1.602000,NaN,NaN
1,Africa Eastern and Southern,AFE,6.650310,6.667308,6.688246,6.709226,6.724930,6.737459,6.766486,6.775493,...,4.615744,4.569884,4.521443,4.471338,4.412973,4.350683,4.287033,4.223820,NaN,NaN
2,Afghanistan,AFG,7.282000,7.284000,7.292000,7.302000,7.304000,7.305000,7.320000,7.339000,...,5.542000,5.433000,5.327000,5.238000,5.145000,5.039000,4.932000,4.840000,NaN,NaN
3,Africa Western and Central,AFW,6.468887,6.478351,6.492277,6.500229,6.516739,6.532766,6.556550,6.582717,...,5.228976,5.098885,4.962571,4.829134,4.707399,4.637741,4.563354,4.497707,NaN,NaN
4,Angola,AGO,6.708000,6.790000,6.872000,6.954000,7.036000,7.116000,7.194000,7.267000,...,5.686000,5.600000,5.519000,5.442000,5.371000,5.304000,5.209000,5.124000,NaN,NaN


In [10]:
# Convertir a formato largo

df_sex_long = df_sex.melt(
    id_vars=["Country Name","Country Code"],
    value_vars=years,
    var_name="year",
    value_name="sex_ratio"
)

df_sex_long["year"] = df_sex_long["year"].astype(int)
df_sex_long.head()

,Country Name,Country Code,year,sex_ratio
0,Aruba,ABW,1960,1.043000
1,Africa Eastern and Southern,AFE,1960,1.033654
2,Afghanistan,AFG,1960,1.049000
3,Africa Western and Central,AFW,1960,1.037556
4,Angola,AGO,1960,1.028000


In [11]:
df_fertility_long = df_fertility.melt(
    id_vars=["Country Name","Country Code"],
    value_vars=years,
    var_name="year",
    value_name="fertility_rate"
)

df_fertility_long["year"] = df_fertility_long["year"].astype(int)
df_fertility_long.head()

,Country Name,Country Code,year,fertility_rate
0,Aruba,ABW,1960,4.567000
1,Africa Eastern and Southern,AFE,1960,6.650310
2,Afghanistan,AFG,1960,7.282000
3,Africa Western and Central,AFW,1960,6.468887
4,Angola,AGO,1960,6.708000


In [12]:
# Crear variables analíticas
df_sex_long['sex_ratio_per_100_female'] = df_sex_long['sex_ratio'] * 100
df_sex_long['deviation_from_105'] = df_sex_long['sex_ratio_per_100_female'] - 105

In [13]:
# Cargar el metadata
ruta_meta = r"C:\Users\GCCM_\OneDrive\Desktop\Bootcamp\proyecto\API_SP.POP.BRTH.MF_DS2_en_csv_v2_8393\Metadata_Country_API_SP.POP.BRTH.MF_DS2_en_csv_v2_8393.csv"
meta = pd.read_csv(ruta_meta)
meta.head()

,Country Code,Region,IncomeGroup,SpecialNotes,TableName,Unnamed: 5
0,ABW,Latin America & Caribbean,High income,NaN,Aruba,NaN
1,AFE,NaN,NaN,"26 countries, stretching from the Red Sea in t...",Africa Eastern and Southern,NaN
2,AFG,Middle East & North Africa,Low income,The reporting period for national accounts dat...,Afghanistan,NaN
3,AFW,NaN,NaN,"22 countries, stretching from the westernmost ...",Africa Western and Central,NaN
4,AGO,Sub-Saharan Africa,Lower middle income,The World Bank systematically assesses the app...,Angola,NaN


In [14]:
# Filtrar solo países reales
meta_paises = meta[meta['Region'].notna() & (meta['Region'] != 'Aggregates')].copy()

iso_paises = set(meta_paises['Country Code'])

In [15]:
# Filtrar países en ambos datasets
df_sex_paises = df_sex_long[df_sex_long["Country Code"].isin(iso_paises)].copy()

df_fertility_paises = df_fertility_long[df_fertility_long["Country Code"].isin(iso_paises)].copy()

In [16]:
# Seleccion de paises
my_50 = [
    "CHN","IND","KOR","VNM","NPL","PAK","BGD","AZE","ARM",
    "ALB","GEO","MNE","SRB","BIH","UZB",
    "USA","IDN","BRA","NGA","MEX","RUS","JPN","ETH","PHL","EGY",
    "TUR","IRN","DEU","THA","FRA","GBR","ITA","ZAF","COL","ARG",
    "KEN","GHA","MAR","TZA","SEN",
    "CHL","PER","CAN","VEN","DOM",
    "ESP","SWE","POL",
    "AUS","NZL"
]

df_sex_50 = df_sex_paises[df_sex_paises['Country Code'].isin(my_50)].copy()


df_fertility_50 = df_fertility_paises[df_fertility_paises["Country Code"].isin(my_50)].copy()

In [17]:
# Eliminar años sin datos reales
df_sex_50 = df_sex_50[df_sex_50["year"] <= 2023]

In [18]:
df_fertility_50 = df_fertility_50[df_fertility_50["year"] <= 2023]

In [19]:
# Unir datasets
df_combined = pd.merge(
    df_sex_50,
    df_fertility_50,
    on=["Country Name","Country Code","year"],
    how="left"
)

df_combined.head()

,Country Name,Country Code,year,sex_ratio,sex_ratio_per_100_female,deviation_from_105,fertility_rate
0,Albania,ALB,1960,1.059,105.9,0.9,6.383
1,Argentina,ARG,1960,1.043,104.3,-0.7,3.136
2,Armenia,ARM,1960,1.064,106.4,1.4,4.786
3,Australia,AUS,1960,1.056,105.6,0.6,3.453
4,Azerbaijan,AZE,1960,1.064,106.4,1.4,5.878


In [20]:
# Crear variable específica para periodos de la política china
def china_policy_phase(country, year):
    if country != 'CHN':
        return 'not_applicable'
    elif year <= 1978:
        return 'pre_policy'
    elif year <= 2015:
        return 'one_child_policy'
    else:
        return 'post_policy'
    
df_combined["china_policy_phase"] = df_combined.apply(
    lambda row: china_policy_phase(row["Country Code"], row["year"]),
    axis=1
)

## Validación del dataset

In [21]:
# Validación final del dataset combinado

# Ordenar dataset
df_combined = df_combined.sort_values(['Country Code', 'year']).reset_index(drop=True)

# Comprobar número de países y dimensiones
df_combined['Country Code'].nunique(), df_combined.shape

# Verificar valores faltantes
df_combined.isna().sum()

# Ver columnas finales
df_combined.columns

df_combined.head()

,Country Name,Country Code,year,sex_ratio,sex_ratio_per_100_female,deviation_from_105,fertility_rate,china_policy_phase
0,Albania,ALB,1960,1.059,105.9,0.9,6.383,not_applicable
1,Albania,ALB,1961,1.059,105.9,0.9,6.273,not_applicable
2,Albania,ALB,1962,1.059,105.9,0.9,6.106,not_applicable
3,Albania,ALB,1963,1.059,105.9,0.9,5.927,not_applicable
4,Albania,ALB,1964,1.059,105.9,0.9,5.714,not_applicable


## Indicadores utilizados

En este análisis se emplean los siguientes indicadores para estudiar la natalidad y el equilibrio en los nacimientos:

- **Número de niños nacidos por cada 100 niñas (sex ratio)**:  
  Indicador principal que permite identificar posibles desequilibrios en los nacimientos. En condiciones naturales, este valor se sitúa en torno a 105.

- **Tasa de fertilidad (fertility rate)**:  
  Representa el número promedio de hijos por mujer y permite analizar la evolución de la natalidad a lo largo del tiempo.

- **Desviación respecto al valor esperado (~105)**:  
  Permite medir en qué medida el sex ratio se aleja del nivel biológico esperado.

Estos indicadores permiten analizar tanto el equilibrio en los nacimientos como la evolución de la natalidad, así como explorar posibles relaciones entre ambos fenómenos.

## Análisis exploratorio de datos (EDA)

En esta sección se realiza un análisis exploratorio de los datos con el objetivo de estudiar la evolución de la natalidad y el equilibrio en los nacimientos, así como identificar posibles relaciones entre ambos fenómenos.

Para ello, se analizan tendencias globales, diferencias entre países y casos específicos que permiten comprender mejor el comportamiento de los indicadores.

#### 1. Análisis global de la fertilidad y el desequilibrio en los nacimientos

In [22]:
# Resumen estadístico general
df_combined.describe().round(2)

,year,sex_ratio,sex_ratio_per_100_female,deviation_from_105,fertility_rate
count,3200.00,3200.00,3200.00,3200.00,3200.00
mean,1991.50,1.06,105.84,0.84,3.37
std,18.48,0.02,2.26,2.26,1.85
min,1960.00,1.01,100.60,-4.40,0.72
25%,1975.75,1.05,104.60,-0.40,1.86
50%,1991.50,1.06,105.50,0.50,2.63
75%,2007.25,1.06,106.40,1.40,4.83
max,2023.00,1.18,117.80,12.80,7.96


In [23]:
# Evolución global conjunta
df_combined.groupby("year")[["fertility_rate","sex_ratio_per_100_female"]] \
.mean().round(2)

,fertility_rate,sex_ratio_per_100_female
year,,
1960,5.16,105.28
1961,5.13,105.28
1962,5.17,105.29
1963,5.16,105.29
1964,5.10,105.31
...,...,...
2019,2.13,105.96
2020,2.08,105.88
2021,2.06,105.82


In [24]:
country_corr = (
    df_combined.groupby("Country Name")[["fertility_rate","sex_ratio_per_100_female"]]
    .corr()
    .unstack()
)

country_corr.head()

fertility_rate                          sex_ratio_per_100_female  \
             fertility_rate sex_ratio_per_100_female           fertility_rate   
Country Name                                                                    
Albania                 1.0                -0.743524                -0.743524   
Argentina               1.0                -0.478246                -0.478246   
Armenia                 1.0                -0.784771                -0.784771   
Australia               1.0                 0.022921                 0.022921   
Azerbaijan              1.0                -0.765742                -0.765742   

                                       
             sex_ratio_per_100_female  
Country Name                           
Albania                           1.0  
Argentina                         1.0  
Armenia                           1.0  
Australia                         1.0  
Azerbaijan                        1.0

Los resultados muestran que la tasa de fertilidad presenta una clara tendencia descendente a nivel global, pasando de valores superiores a 5 hijos por mujer en los años sesenta a valores cercanos a 2 en la actualidad.

Por otro lado, el número de niños nacidos por cada 100 niñas se mantiene relativamente estable en torno al valor biológico esperado (~105), con variaciones moderadas a lo largo del tiempo.

La correlación entre ambas variables es débil, lo que indica que la disminución de la natalidad no se asocia de forma directa con un aumento del desequilibrio en los nacimientos a nivel global.

#### 2. Comparación entre países

In [25]:
# Medias por país 
country_means = (
    df_combined.groupby("Country Name")[["fertility_rate","sex_ratio_per_100_female"]]
    .mean()
    .round(2)
)

# Países con mayor desequilibrio
country_means.sort_values("sex_ratio_per_100_female", ascending=False).head(10)

,fertility_rate,sex_ratio_per_100_female
Country Name,,
China,2.85,111.45
Azerbaijan,3.11,110.24
Armenia,2.41,109.78
Albania,3.17,108.37
Philippines,4.44,108.00
Viet Nam,3.66,108.00
Georgia,2.20,107.86
"Korea, Rep.",2.33,107.85
Montenegro,2.18,107.50


In [26]:
# Países más cercanos al equilibrio
country_means.sort_values("sex_ratio_per_100_female").head(10)

,fertility_rate,sex_ratio_per_100_female
Country Name,,
Kenya,5.97,101.92
Tanzania,6.07,102.83
Senegal,6.07,103.27
Mexico,3.95,103.81
Ghana,5.45,104.01
South Africa,3.83,104.06
Peru,4.08,104.13
Chile,2.61,104.32
Dominican Republic,3.96,104.42


Los resultados muestran diferencias claras entre países en el número de niños nacidos por cada 100 niñas.

Los valores más elevados se concentran principalmente en países de Asia y Europa del Este, como China, Azerbaiyán o Armenia, donde se observan desequilibrios significativos respecto al valor biológico esperado (~105).

Por el contrario, los países más cercanos al equilibrio se encuentran en su mayoría en África y América Latina, como Kenya, Tanzania o Senegal, donde el indicador se mantiene próximo al nivel natural.

Estos resultados indican que el desequilibrio en los nacimientos no es un fenómeno global, sino que se concentra en contextos específicos, lo que sugiere la influencia de factores sociales, culturales o institucionales.

#### 3. Análisis de países representativos

In [27]:
# Selección de países representativos
selected_countries = ["CHN", "KOR", "ESP", "USA", "KEN", "VEN"]

df_selected = df_combined[df_combined["Country Code"].isin(selected_countries)].copy()

# Resumen inicio vs final
summary_selected = (
    df_selected.sort_values("year")
    .groupby("Country Name")
    .agg(
        fertilidad_inicio=("fertility_rate", "first"),
        fertilidad_final=("fertility_rate", "last"),
        sexratio_inicio=("sex_ratio_per_100_female", "first"),
        sexratio_final=("sex_ratio_per_100_female", "last")
    )
    .round(2)
)

# Cálculo de cambios
summary_selected["cambio_fertilidad"] = (
    summary_selected["fertilidad_final"] - summary_selected["fertilidad_inicio"]
)

summary_selected["cambio_sexratio"] = (
    summary_selected["sexratio_final"] - summary_selected["sexratio_inicio"]
)

# Redondear resultados finales
summary_selected = summary_selected.round(2)

summary_selected

,fertilidad_inicio,fertilidad_final,sexratio_inicio,sexratio_final,cambio_fertilidad,cambio_sexratio
Country Name,,,,,,
China,4.45,1.00,106.1,110.8,-3.45,4.7
Kenya,7.65,3.21,101.9,102.2,-4.44,0.3
"Korea, Rep.",5.99,0.72,106.2,105.8,-5.27,-0.4
Spain,2.86,1.12,106.1,106.4,-1.74,0.3
United States,3.65,1.62,104.9,104.9,-2.03,0.0
"Venezuela, RB",6.38,2.08,103.9,105.3,-4.30,1.4


El análisis de países representativos permite identificar patrones diferenciados en la relación entre fertilidad y desequilibrio en los nacimientos.

China muestra una fuerte caída de la fertilidad acompañada de un aumento significativo del sex ratio, lo que indica un desequilibrio creciente en los nacimientos. En contraste, Corea del Sur presenta una reducción de la fertilidad junto con una disminución del desequilibrio, lo que sugiere una corrección progresiva del fenómeno.

En otros países como España o Estados Unidos, la fertilidad también disminuye, pero el sex ratio se mantiene estable en valores cercanos al nivel natural. Por su parte, Kenya presenta una caída importante de la fertilidad sin apenas cambios en el equilibrio de los nacimientos, manteniéndose dentro del rango biológico esperado. De hecho, valores alrededor de 105–106 se consideran normales a nivel internacional 

Estos resultados refuerzan la idea de que la disminución de la natalidad no implica necesariamente un aumento del desequilibrio, sino que depende de factores específicos de cada país.

## Visualización de resultados

In [28]:
# ¿Existe una tendencia global en el desequilibrio de los nacimientos por sexo?

# Promedio global
global_sex = df_combined.groupby("year")["sex_ratio_per_100_female"].mean().reset_index()
global_sex["Country Name"] = "Global"

# Países seleccionados
countries = ["CHN", "KOR", "USA", "VEN", "ESP", "IND","AZE"]

country_data = (
    df_combined[df_combined["Country Code"].isin(countries)]
    [["year", "Country Name", "sex_ratio_per_100_female"]]
    .copy()
)

# Unir datos globales y países
plot_data_global = pd.concat([
    global_sex.rename(columns={"sex_ratio_per_100_female": "value"})[["year", "Country Name", "value"]],
    country_data.rename(columns={"sex_ratio_per_100_female": "value"})
])

# Gráfico
fig = px.line(
    plot_data_global,
    x="year",
    y="value",
    color="Country Name",
    title="Evolución del número de niños nacidos por cada 100 niñas (Global y países seleccionados)",
    labels={
        "year": "Año",
        "value": "Niños por cada 100 niñas",
        "Country Name": "País"
    }
)

fig.add_hline(
    y=105,
    line_dash="dot",
    line_color="gray",
    opacity=0.6,
    annotation_text="Valor esperado (~105)",
    annotation_position="top right"
)

fig.write_html("grafico_1_sexratio_global_paises.html")
fig.show()

In [29]:
# ¿Está cayendo la natalidad a nivel mundial? 

global_fert = df_combined.groupby("year")["fertility_rate"].mean().reset_index() 

fig = px.line( 
    global_fert, 
    x="year", 
    y="fertility_rate", 
    title="Evolución global de la tasa de fertilidad", 
    labels={ "year": "Año", "fertility_rate": "Hijos por mujer" 
        } 
    ) 

fig.write_html("grafico_2_fertilidad_global.html") 
fig.show()

In [30]:
# ¿Qué países presentan desequilibrios en los nacimientos y cuáles se mantienen cerca del nivel biológico esperado?

# Lista de países
countries = ["CHN", "KOR", "USA", "VEN", "ESP", "IND", "AZE", "KEN"]

# Filtrar + agrupar
plot_data_points = (
    df_combined[df_combined["Country Code"].isin(countries)]
    .groupby(["Country Name", "Country Code"], as_index=False)["sex_ratio_per_100_female"]
    .mean()
)

# Calcular desviación respecto a 105
plot_data_points["desviacion"] = abs(
    plot_data_points["sex_ratio_per_100_female"] - 105
)

# Tamaño con base mínima visible
plot_data_points["size"] = plot_data_points["desviacion"] * 2 + 6

fig = px.scatter(
    plot_data_points,
    x="sex_ratio_per_100_female",
    y="Country Name",
    size="size",
    size_max=24,
    hover_data={
        "size": False 
    },
    title="Comparación del número de niños nacidos por cada 100 niñas por país",
    labels={
        "sex_ratio_per_100_female": "Niños por cada 100 niñas",
        "Country Name": "País"
    }
)

fig.add_vline(
    x=105,
    line_dash="dash",
    annotation_text="Valor esperado (~105)"
)

fig.update_traces(
    marker=dict(
        line=dict(width=1.5, color="darkblue"),
        opacity=0.85
    )
)

fig.update_layout(
    margin=dict(l=160, r=40, t=60, b=60)
)

fig.write_html("grafico_3_puntos_desequilibrio.html")
fig.show()

In [31]:
# ¿Qué países han aumentado más su sex ratio con el tiempo?
change = (
    df_combined.sort_values("year")
    .groupby("Country Name")
    .agg(
        inicio=("sex_ratio_per_100_female", "first"),
        fin=("sex_ratio_per_100_female", "last")
    )
)

change["cambio"] = change["fin"] - change["inicio"]

# Top países con mayor aumento
top_change = (
    change.sort_values("cambio", ascending=False)
    .head(10)
    .reset_index()
)

import plotly.express as px

fig = px.bar(
    top_change,
    x="cambio",
    y="Country Name",
    orientation="h",
    title="Países con mayor aumento del número de niños por cada 100 niñas (1960–2023)",
    labels={
        "cambio": "Cambio en niños por cada 100 niñas",
        "Country Name": "País"
    }
)

fig.write_html("grafico_4_extra_cambio_temporal.html")
fig.show()

In [32]:
# ¿Cómo ha evolucionado el desequilibrio en los nacimientos por sexo en países seleccionados a lo largo del tiempo?

countries = ["CHN", "KOR", "ESP", "USA", "VEN","AZE","IND"]

compare_sex = df_combined[df_combined["Country Code"].isin(countries)].copy()

fig = px.line(
    compare_sex,
    x="year",
    y="sex_ratio_per_100_female",
    color="Country Name",
    title="Evolución del número de niños nacidos por cada 100 niñas en países seleccionados",
    labels={
        "year": "Año",
        "sex_ratio_per_100_female": "Niños por cada 100 niñas",
        "Country Name": "País"
    }
)

fig.update_yaxes(range=[103, 119])

fig.add_hline(
    y=105,
    line_dash="dot",
    line_color="gray",
    opacity=0.6,
    annotation_text="Valor esperado (~105)",
    annotation_position="bottom right"
)

fig.write_html("grafico_5_comparacion_paises_sexratio.html")
fig.show()

In [33]:
# ¿Existe una relación entre la tasa de fertilidad y el desequilibrio en los nacimientos?

country_means = (
    df_combined.groupby("Country Name")[["fertility_rate", "sex_ratio_per_100_female"]]
    .mean()
    .reset_index()
)


highlight_countries = ["China", "Korea, Rep.", "India", "Azerbaijan", "Armenia", "Kenya","Venezuela, RB","Spain"]

country_means["label"] = country_means["Country Name"].where(
    country_means["Country Name"].isin(highlight_countries),
    ""
)

fig = px.scatter(
    country_means,
    x="fertility_rate",
    y="sex_ratio_per_100_female",
    text="label",
    trendline="ols",
    title="Relación entre fertilidad y número de niños nacidos por cada 100 niñas",
    labels={
        "fertility_rate": "Hijos por mujer",
        "sex_ratio_per_100_female": "Niños por cada 100 niñas"
    },
    hover_name="Country Name"
)

# Línea horizontal de referencia
fig.add_hline(
    y=105,
    line_dash="dot",
    line_color="gray",
    opacity=0.6,
    annotation_text="Valor esperado (~105)",
    annotation_position="top right"
)

# Ajustar posición del texto
fig.update_traces(
    textposition="top center",
    marker=dict(size=9)
)

fig.write_html("grafico_6_scatter_fertilidad_sexratio_regresion.html")
fig.show()

In [34]:
# ¿Está la caída de la natalidad asociada a un desequilibrio en los nacimientos a nivel global?

global_both = (
    df_combined.groupby("year")[["fertility_rate", "sex_ratio_per_100_female"]]
    .mean()
    .reset_index()
)

fig = go.Figure()

# Fertilidad
fig.add_trace(go.Scatter(
    x=global_both["year"],
    y=global_both["fertility_rate"],
    mode="lines",
    name="Fertilidad"
))

# Sex ratio
fig.add_trace(go.Scatter(
    x=global_both["year"],
    y=global_both["sex_ratio_per_100_female"],
    mode="lines",
    name="Niños por cada 100 niñas",
    yaxis="y2"
))

# Layout 
fig.update_layout(
    title="Evolución global de la fertilidad y del número de niños nacidos por cada 100 niñas",
    xaxis_title="Año",
    
    yaxis=dict(
        title="Hijos por mujer",
        tickmode="linear",
        dtick=0.5
    ),
    
    yaxis2=dict(
        title="Niños por cada 100 niñas",
        overlaying="y",
        side="right",
        range=[103, 108],  
        tickmode="linear",
        dtick=1
    )
)

# Línea de referencia (105)
fig.add_shape(
    type="line",
    x0=global_both["year"].min(),
    x1=global_both["year"].max(),
    y0=105,
    y1=105,
    xref="x",
    yref="y2",
    line=dict(color="gray", dash="dot")
)

# Texto de la línea
fig.add_annotation(
    x=global_both["year"].max(),
    y=105,
    xref="x",
    yref="y2",
    text="Valor esperado (~105)",
    showarrow=False,
    xanchor="right",
    yanchor="bottom"
)

fig.write_html("grafico_7_global_conjunto.html")
fig.show()

In [35]:
# ¿Coincide la caída de fertilidad con el aumento del desequilibrio?

countries = ["CHN", "KOR", "ESP", "USA", "VEN","AZE","IND"]

country_names = {
    "CHN": "China",
    "KOR": "Corea del Sur",
    "ESP": "España",
    "USA": "Estados Unidos",
    "VEN": "Venezuela",
    "AZE": "Azerbaijan",
    "IND": "India"
}

# crear figura
fig = go.Figure()

# añadir todas las trazas pero ocultas
for code in countries:
    data = (
        df_combined[df_combined["Country Code"] == code]
        .groupby("year")[["fertility_rate", "sex_ratio_per_100_female"]]
        .mean()
        .reset_index()
    )

    # fertilidad
    fig.add_trace(go.Scatter(
        x=data["year"],
        y=data["fertility_rate"],
        mode="lines",
        name=f"{country_names[code]} - Fertilidad",
        visible=(code == "CHN")  # solo China visible al inicio
    ))

    # sex ratio
    fig.add_trace(go.Scatter(
        x=data["year"],
        y=data["sex_ratio_per_100_female"],
        mode="lines",
        name=f"{country_names[code]} - Sex ratio",
        yaxis="y2",
        visible=(code == "CHN")
    ))

# crear botones dropdown
buttons = []

for i, code in enumerate(countries):
    visibility = [False] * (len(countries) * 2)
    
    # activar solo las dos trazas del país seleccionado
    visibility[i*2] = True
    visibility[i*2 + 1] = True

    buttons.append(dict(
        label=country_names[code],
        method="update",
        args=[{"visible": visibility}]
    ))

fig.update_layout(
    title="Relación entre fertilidad y desequilibrio por país",
    xaxis_title="Año",
    yaxis=dict(title="Hijos por mujer"),
    yaxis2=dict(
        title="Niños por cada 100 niñas",
        overlaying="y",
        side="right",
        range=[101, 119]
    ),
    updatemenus=[
        dict(
            buttons=buttons,
            direction="down",
            showactive=True,
            x=0.1,
            y=1.15
        )
    ]
)

fig.write_html("grafico_8_dropdown_paises.html")
fig.show()

In [36]:
# ¿Cómo se distribuye el desequilibrio geográficamente?

map_data = (
    df_combined.groupby(["Country Name", "Country Code"])["sex_ratio_per_100_female"]
    .mean()
    .reset_index()
)

fig = px.choropleth(
    map_data,
    locations="Country Code",
    color="sex_ratio_per_100_female",
    hover_name="Country Name",
    title="Distribución geográfica del número de niños nacidos por cada 100 niñas",
    labels={"sex_ratio_per_100_female": "Niños por cada 100 niñas"}
)

fig.write_html("grafico_9_mapa_sexratio.html")
fig.show()

## Visualización interactiva (Flourish)

Con el objetivo de representar la evolución temporal de los datos de forma dinámica, se generan datasets para visualizaciones en Flourish.

In [37]:
# 1. Evolución del desequilibrio en los nacimientos por país

# Países seleccionados
countries = [
    "China", "Korea, Rep.", "India",
    "United States", "Spain", "Kenya", "Venezuela, RB","Azerbaijan"
]

# Filtrar
df_flourish = df_combined[df_combined["Country Name"].isin(countries)].copy()

# Cambiar nombres conflictivos
df_flourish["Country Name"] = df_flourish["Country Name"].replace({
    "Korea, Rep.": "South Korea",
    "Venezuela, RB": "Venezuela"
})

# Pivot a formato ancho
flourish_race = df_flourish.pivot(
    index="Country Name",
    columns="year",
    values="sex_ratio_per_100_female"
).reset_index()

# Renombrar columna principal
flourish_race = flourish_race.rename(columns={"Country Name": "Country"})

# Redondear para que quede limpio
year_cols = [col for col in flourish_race.columns if col != "Country"]
flourish_race[year_cols] = flourish_race[year_cols].round(1)

# Exportar a Excel
flourish_race.to_excel("flourish_sexratio_race.xlsx", index=False)

flourish_race.head()

year,Country,1960,1961,1962,1963,1964,1965,1966,1967,1968,...,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023
0,Azerbaijan,106.4,106.4,106.4,106.4,106.4,106.3,106.3,106.3,106.3,...,115.1,114.6,114.3,114.0,113.8,113.4,113.0,112.6,112.2,111.8
1,China,106.1,106.1,106.2,106.2,106.2,106.2,106.2,106.3,106.3,...,115.5,115.0,114.5,113.3,113.9,113.9,111.2,110.4,110.6,110.8
2,India,105.2,105.2,105.2,105.2,105.2,105.2,105.3,105.3,105.3,...,109.1,109.0,108.8,108.6,108.4,108.2,107.9,107.7,107.5,107.3
3,Kenya,101.9,102.0,101.9,101.9,102.0,102.0,102.0,102.0,102.0,...,102.2,102.2,102.2,102.2,102.2,102.2,102.2,102.2,102.2,102.2
4,South Korea,106.2,106.2,106.2,106.2,106.2,106.2,106.2,106.2,106.2,...,105.3,105.3,105.2,106.1,105.5,105.6,105.7,105.7,105.8,105.8


In [38]:
# 2. Relación entre fertilidad y desequilibrio en los nacimientos

# Dataset para burbujas
flourish_bubble = (
    df_combined[df_combined["Country Name"].isin(countries)]
    [["year", "Country Name", "fertility_rate", "sex_ratio_per_100_female"]]
    .copy()
)

# Renombrar columnas
flourish_bubble = flourish_bubble.rename(columns={
    "year": "Year",
    "Country Name": "Country",
    "fertility_rate": "Fertility",
    "sex_ratio_per_100_female": "Sex Ratio"
})

# tamaño de burbuja fijo
flourish_bubble["Size"] = 1

# Exportar
flourish_bubble.to_csv("flourish_bubble_fertility_sexratio.csv", index=False)

flourish_bubble.head()

,Year,Country,Fertility,Sex Ratio,Size
256,1960,Azerbaijan,5.878,106.4,1
257,1961,Azerbaijan,5.937,106.4,1
258,1962,Azerbaijan,5.961,106.4,1
259,1963,Azerbaijan,5.945,106.4,1
260,1964,Azerbaijan,5.888,106.4,1


## Exportación del dataset final

In [39]:
# Exportar dataset final limpio para Power BI
df_combined.to_csv(r"C:\Users\GCCM_\OneDrive\Desktop\Bootcamp\proyecto\sex_ratio_dataset_final.csv", index=False)